File: app/src/main/java/me/cargoapp/cargo/helper/PermissionHelper.java
package me.cargoapp.cargo.helper;

import android.Manifest;
import android.content.ComponentName;
import android.content.Context;
import android.content.pm.PackageManager;
import android.provider.Settings;
import android.text.TextUtils;

public class PermissionHelper {
    private String TAG = this.getClass().getSimpleName();

    public static boolean isPermittedTo(Context context, String permission) {
        switch (permission) {
            case Manifest.permission.SYSTEM_ALERT_WINDOW:
                return Settings.canDrawOverlays(context);
            case Manifest.permission.BIND_NOTIFICATION_LISTENER_SERVICE:
                String pkgName = context.getPackageName();
                final String flat = Settings.Secure.getString(context.getContentResolver(), "enabled_notification_listeners");
                if (TextUtils.isEmpty(flat)) return false;

                final String[] names = flat.split(":");
                for (String name : names) {
                    final ComponentName cn = ComponentName.unflattenFromString(name);
                    if (cn == null) continue;

                    if (TextUtils.equals(pkgName, cn.getPackageName())) {
                        return true;
                    }
                }
                return false;
            default:
                return context.checkSelfPermission(permission) == PackageManager.PERMISSION_GRANTED;
        }
    }

    public static boolean isPermittedTo(Context context, String[] permissions) {
        boolean permissionsOk = true;

        for (String permission : permissions) {
            if (!PermissionHelper.isPermittedTo(context, permission)) {
                permissionsOk = false;
                break;
            }
        }

        return permissionsOk;
    }
}

File: app/src/main/java/me/cargoapp/cargo/helper/PermissionHelper.kt
package me.cargoapp.cargo.helper

import android.Manifest
import android.content.ComponentName
import android.content.Context
import android.content.pm.PackageManager
import android.provider.Settings
import android.text.TextUtils

object PermissionHelper {

    fun isPermittedTo(context: Context, permission: String): Boolean {
        when (permission) {
            Manifest.permission.SYSTEM_ALERT_WINDOW -> return Settings.canDrawOverlays(context)
            Manifest.permission.BIND_NOTIFICATION_LISTENER_SERVICE -> {
                val pkgName = context.packageName
                val flat = Settings.Secure.getString(context.contentResolver, "enabled_notification_listeners")
                if (TextUtils.isEmpty(flat)) return false

                val names = flat.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                for (name in names) {
                    val cn = ComponentName.unflattenFromString(name) ?: continue

                    if (TextUtils.equals(pkgName, cn.packageName)) {
                        return true
                    }
                }
                return false
            }
            else -> return context.checkSelfPermission(permission) == PackageManager.PERMISSION_GRANTED
        }
    }

    fun isPermittedTo(context: Context, permissions: Array<String>): Boolean {
        var permissionsOk = true

        for (permission in permissions) {
            if (!PermissionHelper.isPermittedTo(context, permission)) {
                permissionsOk = false
                break
            }
        }

        return permissionsOk
    }
}